In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
data = pd.read_csv('/content/drive/MyDrive/datasets/Data_Science_Internship.csv')
data.head()

Unnamed: 0                                           Agent_id status  \
0           0  1deba9e96f404694373de9749ddd1ca8aa7bb823145a6f...   LOST   
1           1  299ae77a4ef350ae0dd37d6bba1c002d03444fb1edb236...   LOST   
2           2  c213697430c006013012dd2aca82dd9732aa0a1a6bca13...   LOST   
3           3  eac9815a500f908736d303e23aa227f0957177b0e6756b...   LOST   
4           4  1deba9e96f404694373de9749ddd1ca8aa7bb823145a6f...   LOST   

      lost_reason                budget                                lease  \
0  Not responding                   NaN                                  NaN   
1      Low budget                   NaN                                  NaN   
2  Not responding  £121 - £180 Per Week  Full Year Course Stay 40 - 44 weeks   
3      Low budget                   0-0                                    0   
4       Junk lead                   NaN                                  NaN   

     movein                                             source  \
0       NaN  9b2d5b4678781e53038e91ea5324530a03f27dc1d0e5f6...   
1       NaN  9b2d5b4678781e53038e91ea5324530a03f27dc1d0e5f6...   
2  31/08/22  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   
3       NaN  ba2d0a29556ac20f86f45e4543c0825428cba33fd7a9ea...   
4       NaN  9b2d5b4678781e53038e91ea5324530a03f27dc1d0e5f6...   

                                         source_city  \
0  9b2d5b4678781e53038e91ea5324530a03f27dc1d0e5f6...   
1  9b2d5b4678781e53038e91ea5324530a03f27dc1d0e5f6...   
2  9b8cc3c63cdf447e463c11544924bf027945cbd29675f7...   
3  a5f0d2d08eb0592087e3a3a2f9c1ba2c67cc30f2efd2bd...   
4  9b2d5b4678781e53038e91ea5324530a03f27dc1d0e5f6...   

                                      source_country  \
0  9b2d5b4678781e53038e91ea5324530a03f27dc1d0e5f6...   
1  9b2d5b4678781e53038e91ea5324530a03f27dc1d0e5f6...   
2  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
3  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
4  9b2d5b4678781e53038e91ea5324530a03f27dc1d0e5f6...   

                                          utm_source  \
0  3d59f7548e1af2151b64135003ce63c0a484c26b9b8b16...   
1  3d59f7548e1af2151b64135003ce63c0a484c26b9b8b16...   
2  bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07...   
3  bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07...   
4  3d59f7548e1af2151b64135003ce63c0a484c26b9b8b16...   

                                          utm_medium  \
0  268ad70eb5bc4737a2ae28162cbca30118cc94520e49ef...   
1  268ad70eb5bc4737a2ae28162cbca30118cc94520e49ef...   
2  09076eb7665d1fb9389c7c4517fee0b00e43092eb34821...   
3  09076eb7665d1fb9389c7c4517fee0b00e43092eb34821...   
4  268ad70eb5bc4737a2ae28162cbca30118cc94520e49ef...   

                                            des_city  \
0  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   
1  5372372f3bf5896820cb2819300c3e681820d82c6efc54...   
2  11ab03a1a8c367191355c152f39fe28cae5e426fce49ef...   
3  19372fa44c57a01c37a5a8418779ca3d99b0b59731fb35...   
4  9b2d5b4678781e53038e91ea5324530a03f27dc1d0e5f6...   

                                         des_country room_type  \
0  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...       NaN   
1  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...       NaN   
2  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...   Ensuite   
3  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...       NaN   
4  9b2d5b4678781e53038e91ea5324530a03f27dc1d0e5f6...       NaN   

                                             lead_id  
0  cd5dc0d9393f3980d11d4ba6f88f8110c2b7a7f7796307...  
1  b94693673a5f7178d1b114e4004ad52377d3244dd24a3d...  
2  96ea4e2bf04496c044745938c0299c264c3f4ba079e572...  
3  1d2b34d8add02a182a4129023766ca4585a8ddced0e5b3...  
4  fc10fffd29cfbe93c55158fb47752a7501c211d253468c...

## Data Cleaning

In [3]:
data.shape

(46608, 16)

In [4]:
data.columns

Index(['Unnamed: 0', 'Agent_id', 'status', 'lost_reason', 'budget', 'lease',
       'movein', 'source', 'source_city', 'source_country', 'utm_source',
       'utm_medium', 'des_city', 'des_country', 'room_type', 'lead_id'],
      dtype='object')

In [5]:
data.drop(['Unnamed: 0'], axis=1, inplace = True)

In [6]:
data = data.replace('9b2d5b4678781e53038e91ea5324530a03f27dc1d0e5f6c9bc9d493a23be9de0', np.nan)

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46608 entries, 0 to 46607
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Agent_id        46608 non-null  object
 1   status          46608 non-null  object
 2   lost_reason     43244 non-null  object
 3   budget          42908 non-null  object
 4   lease           44267 non-null  object
 5   movein          32970 non-null  object
 6   source          40631 non-null  object
 7   source_city     37757 non-null  object
 8   source_country  37967 non-null  object
 9   utm_source      46547 non-null  object
 10  utm_medium      43421 non-null  object
 11  des_city        44071 non-null  object
 12  des_country     44071 non-null  object
 13  room_type       23061 non-null  object
 14  lead_id         46608 non-null  object
dtypes: object(15)
memory usage: 5.3+ MB


In [8]:
data.describe().T

count unique  \
Agent_id        46608    120   
status          46608      6   
lost_reason     43244     30   
budget          42908   1857   
lease           44267    311   
movein          32970    477   
source          40631    682   
source_city     37757   4335   
source_country  37967    185   
utm_source      46547     34   
utm_medium      43421     63   
des_city        44071    219   
des_country     44071     14   
room_type       23061      5   
lead_id         46608  30574   

                                                              top   freq  
Agent_id        2fca346db656187102ce806ac732e06a62df0dbb2829e5...   8186  
status                                                       LOST  43244  
lost_reason                                      Low availability   8892  
budget                                        £60 - £120 Per week   9862  
lease                         Full Year Course Stay 40 - 44 weeks  11111  
movein                                                   10/09/22   2453  
source          7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...  19306  
source_city     ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   3044  
source_country  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...  18812  
utm_source      bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07...  31125  
utm_medium      09076eb7665d1fb9389c7c4517fee0b00e43092eb34821...  31761  
des_city        ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   7361  
des_country     8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...  42674  
room_type                                                 Ensuite  12477  
lead_id         e741bdc636f2e99f2f83b31a64ff4b6d07f2b6a182d1cc...      2

In [9]:
# Find missing value percentage for each column

def missing_percentage():
  cols = data.columns
  missing = dict()

  for c in cols:
    missing[c] = (data[c].isna().sum() / data.shape[0])*100
  missing_df = pd.DataFrame({'column': missing.keys(), 'missing_percentage': missing.values()})
  return missing_df

In [10]:
missing_percentage()

column  missing_percentage
0         Agent_id            0.000000
1           status            0.000000
2      lost_reason            7.217645
3           budget            7.938551
4            lease            5.022743
5           movein           29.261071
6           source           12.823979
7      source_city           18.990302
8   source_country           18.539736
9       utm_source            0.130879
10      utm_medium            6.837882
11        des_city            5.443272
12     des_country            5.443272
13       room_type           50.521370
14         lead_id            0.000000

In [11]:
# drop room_type as more that 50% missing
data = data.drop('room_type', axis=1)

In [12]:
data.head()

Agent_id status     lost_reason  \
0  1deba9e96f404694373de9749ddd1ca8aa7bb823145a6f...   LOST  Not responding   
1  299ae77a4ef350ae0dd37d6bba1c002d03444fb1edb236...   LOST      Low budget   
2  c213697430c006013012dd2aca82dd9732aa0a1a6bca13...   LOST  Not responding   
3  eac9815a500f908736d303e23aa227f0957177b0e6756b...   LOST      Low budget   
4  1deba9e96f404694373de9749ddd1ca8aa7bb823145a6f...   LOST       Junk lead   

                 budget                                lease    movein  \
0                   NaN                                  NaN       NaN   
1                   NaN                                  NaN       NaN   
2  £121 - £180 Per Week  Full Year Course Stay 40 - 44 weeks  31/08/22   
3                   0-0                                    0       NaN   
4                   NaN                                  NaN       NaN   

                                              source  \
0                                                NaN   
1                                                NaN   
2  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   
3  ba2d0a29556ac20f86f45e4543c0825428cba33fd7a9ea...   
4                                                NaN   

                                         source_city  \
0                                                NaN   
1                                                NaN   
2  9b8cc3c63cdf447e463c11544924bf027945cbd29675f7...   
3  a5f0d2d08eb0592087e3a3a2f9c1ba2c67cc30f2efd2bd...   
4                                                NaN   

                                      source_country  \
0                                                NaN   
1                                                NaN   
2  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
3  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
4                                                NaN   

                                          utm_source  \
0  3d59f7548e1af2151b64135003ce63c0a484c26b9b8b16...   
1  3d59f7548e1af2151b64135003ce63c0a484c26b9b8b16...   
2  bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07...   
3  bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07...   
4  3d59f7548e1af2151b64135003ce63c0a484c26b9b8b16...   

                                          utm_medium  \
0  268ad70eb5bc4737a2ae28162cbca30118cc94520e49ef...   
1  268ad70eb5bc4737a2ae28162cbca30118cc94520e49ef...   
2  09076eb7665d1fb9389c7c4517fee0b00e43092eb34821...   
3  09076eb7665d1fb9389c7c4517fee0b00e43092eb34821...   
4  268ad70eb5bc4737a2ae28162cbca30118cc94520e49ef...   

                                            des_city  \
0  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   
1  5372372f3bf5896820cb2819300c3e681820d82c6efc54...   
2  11ab03a1a8c367191355c152f39fe28cae5e426fce49ef...   
3  19372fa44c57a01c37a5a8418779ca3d99b0b59731fb35...   
4                                                NaN   

                                         des_country  \
0  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...   
1  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...   
2  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...   
3  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...   
4                                                NaN   

                                             lead_id  
0  cd5dc0d9393f3980d11d4ba6f88f8110c2b7a7f7796307...  
1  b94693673a5f7178d1b114e4004ad52377d3244dd24a3d...  
2  96ea4e2bf04496c044745938c0299c264c3f4ba079e572...  
3  1d2b34d8add02a182a4129023766ca4585a8ddced0e5b3...  
4  fc10fffd29cfbe93c55158fb47752a7501c211d253468c...

In [13]:
missing_percentage()

column  missing_percentage
0         Agent_id            0.000000
1           status            0.000000
2      lost_reason            7.217645
3           budget            7.938551
4            lease            5.022743
5           movein           29.261071
6           source           12.823979
7      source_city           18.990302
8   source_country           18.539736
9       utm_source            0.130879
10      utm_medium            6.837882
11        des_city            5.443272
12     des_country            5.443272
13         lead_id            0.000000

In [14]:
data.describe().T

count unique  \
Agent_id        46608    120   
status          46608      6   
lost_reason     43244     30   
budget          42908   1857   
lease           44267    311   
movein          32970    477   
source          40631    682   
source_city     37757   4335   
source_country  37967    185   
utm_source      46547     34   
utm_medium      43421     63   
des_city        44071    219   
des_country     44071     14   
lead_id         46608  30574   

                                                              top   freq  
Agent_id        2fca346db656187102ce806ac732e06a62df0dbb2829e5...   8186  
status                                                       LOST  43244  
lost_reason                                      Low availability   8892  
budget                                        £60 - £120 Per week   9862  
lease                         Full Year Course Stay 40 - 44 weeks  11111  
movein                                                   10/09/22   2453  
source          7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...  19306  
source_city     ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   3044  
source_country  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...  18812  
utm_source      bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07...  31125  
utm_medium      09076eb7665d1fb9389c7c4517fee0b00e43092eb34821...  31761  
des_city        ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   7361  
des_country     8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...  42674  
lead_id         e741bdc636f2e99f2f83b31a64ff4b6d07f2b6a182d1cc...      2

In [15]:
#replacing missing values with mode
from sklearn.impute import SimpleImputer

cols = data.columns

data = pd.DataFrame(SimpleImputer(strategy='most_frequent').fit_transform(data))

data.columns = cols

data.head()


Agent_id status     lost_reason  \
0  1deba9e96f404694373de9749ddd1ca8aa7bb823145a6f...   LOST  Not responding   
1  299ae77a4ef350ae0dd37d6bba1c002d03444fb1edb236...   LOST      Low budget   
2  c213697430c006013012dd2aca82dd9732aa0a1a6bca13...   LOST  Not responding   
3  eac9815a500f908736d303e23aa227f0957177b0e6756b...   LOST      Low budget   
4  1deba9e96f404694373de9749ddd1ca8aa7bb823145a6f...   LOST       Junk lead   

                 budget                                lease    movein  \
0   £60 - £120 Per week  Full Year Course Stay 40 - 44 weeks  10/09/22   
1   £60 - £120 Per week  Full Year Course Stay 40 - 44 weeks  10/09/22   
2  £121 - £180 Per Week  Full Year Course Stay 40 - 44 weeks  31/08/22   
3                   0-0                                    0  10/09/22   
4   £60 - £120 Per week  Full Year Course Stay 40 - 44 weeks  10/09/22   

                                              source  \
0  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   
1  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   
2  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   
3  ba2d0a29556ac20f86f45e4543c0825428cba33fd7a9ea...   
4  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   

                                         source_city  \
0  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   
1  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   
2  9b8cc3c63cdf447e463c11544924bf027945cbd29675f7...   
3  a5f0d2d08eb0592087e3a3a2f9c1ba2c67cc30f2efd2bd...   
4  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   

                                      source_country  \
0  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
1  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
2  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
3  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
4  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   

                                          utm_source  \
0  3d59f7548e1af2151b64135003ce63c0a484c26b9b8b16...   
1  3d59f7548e1af2151b64135003ce63c0a484c26b9b8b16...   
2  bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07...   
3  bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07...   
4  3d59f7548e1af2151b64135003ce63c0a484c26b9b8b16...   

                                          utm_medium  \
0  268ad70eb5bc4737a2ae28162cbca30118cc94520e49ef...   
1  268ad70eb5bc4737a2ae28162cbca30118cc94520e49ef...   
2  09076eb7665d1fb9389c7c4517fee0b00e43092eb34821...   
3  09076eb7665d1fb9389c7c4517fee0b00e43092eb34821...   
4  268ad70eb5bc4737a2ae28162cbca30118cc94520e49ef...   

                                            des_city  \
0  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   
1  5372372f3bf5896820cb2819300c3e681820d82c6efc54...   
2  11ab03a1a8c367191355c152f39fe28cae5e426fce49ef...   
3  19372fa44c57a01c37a5a8418779ca3d99b0b59731fb35...   
4  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   

                                         des_country  \
0  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...   
1  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...   
2  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...   
3  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...   
4  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...   

                                             lead_id  
0  cd5dc0d9393f3980d11d4ba6f88f8110c2b7a7f7796307...  
1  b94693673a5f7178d1b114e4004ad52377d3244dd24a3d...  
2  96ea4e2bf04496c044745938c0299c264c3f4ba079e572...  
3  1d2b34d8add02a182a4129023766ca4585a8ddced0e5b3...  
4  fc10fffd29cfbe93c55158fb47752a7501c211d253468c...

In [16]:
missing_percentage()

column  missing_percentage
0         Agent_id                 0.0
1           status                 0.0
2      lost_reason                 0.0
3           budget                 0.0
4            lease                 0.0
5           movein                 0.0
6           source                 0.0
7      source_city                 0.0
8   source_country                 0.0
9       utm_source                 0.0
10      utm_medium                 0.0
11        des_city                 0.0
12     des_country                 0.0
13         lead_id                 0.0

In [17]:
data.head()

Agent_id status     lost_reason  \
0  1deba9e96f404694373de9749ddd1ca8aa7bb823145a6f...   LOST  Not responding   
1  299ae77a4ef350ae0dd37d6bba1c002d03444fb1edb236...   LOST      Low budget   
2  c213697430c006013012dd2aca82dd9732aa0a1a6bca13...   LOST  Not responding   
3  eac9815a500f908736d303e23aa227f0957177b0e6756b...   LOST      Low budget   
4  1deba9e96f404694373de9749ddd1ca8aa7bb823145a6f...   LOST       Junk lead   

                 budget                                lease    movein  \
0   £60 - £120 Per week  Full Year Course Stay 40 - 44 weeks  10/09/22   
1   £60 - £120 Per week  Full Year Course Stay 40 - 44 weeks  10/09/22   
2  £121 - £180 Per Week  Full Year Course Stay 40 - 44 weeks  31/08/22   
3                   0-0                                    0  10/09/22   
4   £60 - £120 Per week  Full Year Course Stay 40 - 44 weeks  10/09/22   

                                              source  \
0  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   
1  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   
2  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   
3  ba2d0a29556ac20f86f45e4543c0825428cba33fd7a9ea...   
4  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   

                                         source_city  \
0  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   
1  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   
2  9b8cc3c63cdf447e463c11544924bf027945cbd29675f7...   
3  a5f0d2d08eb0592087e3a3a2f9c1ba2c67cc30f2efd2bd...   
4  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   

                                      source_country  \
0  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
1  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
2  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
3  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
4  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   

                                          utm_source  \
0  3d59f7548e1af2151b64135003ce63c0a484c26b9b8b16...   
1  3d59f7548e1af2151b64135003ce63c0a484c26b9b8b16...   
2  bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07...   
3  bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07...   
4  3d59f7548e1af2151b64135003ce63c0a484c26b9b8b16...   

                                          utm_medium  \
0  268ad70eb5bc4737a2ae28162cbca30118cc94520e49ef...   
1  268ad70eb5bc4737a2ae28162cbca30118cc94520e49ef...   
2  09076eb7665d1fb9389c7c4517fee0b00e43092eb34821...   
3  09076eb7665d1fb9389c7c4517fee0b00e43092eb34821...   
4  268ad70eb5bc4737a2ae28162cbca30118cc94520e49ef...   

                                            des_city  \
0  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   
1  5372372f3bf5896820cb2819300c3e681820d82c6efc54...   
2  11ab03a1a8c367191355c152f39fe28cae5e426fce49ef...   
3  19372fa44c57a01c37a5a8418779ca3d99b0b59731fb35...   
4  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   

                                         des_country  \
0  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...   
1  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...   
2  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...   
3  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...   
4  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...   

                                             lead_id  
0  cd5dc0d9393f3980d11d4ba6f88f8110c2b7a7f7796307...  
1  b94693673a5f7178d1b114e4004ad52377d3244dd24a3d...  
2  96ea4e2bf04496c044745938c0299c264c3f4ba079e572...  
3  1d2b34d8add02a182a4129023766ca4585a8ddced0e5b3...  
4  fc10fffd29cfbe93c55158fb47752a7501c211d253468c...

In [18]:
data.describe().T['unique']

Agent_id            120
status                6
lost_reason          30
budget             1857
lease               311
movein              477
source              682
source_city        4335
source_country      185
utm_source           34
utm_medium           63
des_city            219
des_country          14
lead_id           30574
Name: unique, dtype: object

In [19]:
data.shape

(46608, 14)

In [20]:
data['status'].value_counts()

LOST           43244
WON             3073
OPPORTUNITY      254
CONTACTED         28
PROCESSING         8
IMPORTANT          1
Name: status, dtype: int64

In [21]:
# We can remove all entries except LOST and WON

data = data[(data['status'] == 'WON') | (data['status'] == 'LOST')]

In [22]:
data['status'].value_counts()

LOST    43244
WON      3073
Name: status, dtype: int64

In [23]:
# Printing Normalized value counts of all columns

def value_count_all(data):
  for c in data.columns:
    print(data[c].value_counts(normalize=True)*100)

value_count_all(data)

2fca346db656187102ce806ac732e06a62df0dbb2829e511a770556d398e1a6e    17.673856
96ab6152cb421b4d4012a4903c19004608a342a5f7c6671a7828b8ad8d5b9489     3.132759
44864c96fa1c36602f0d045b268981b6cab638a60fc207c89e3bba235f619450     2.761405
fb3e2c0015ad5bcd60d7619ca575a56c15c5d2cd3f01f18e68c4dfaeee585db5     2.385733
eac9815a500f908736d303e23aa227f0957177b0e6756b915a3d3cc3a846b83b     2.308008
                                                                      ...    
e16d12cd21f38bc8253f65cc5c0bf4932cf1397ba1c60682b2a99049e0e7ceed     0.002159
04222ea3e14cb1209b9726defe3efce5196b7afa0a959854a30401be41f4026d     0.002159
fb87e6cf60af08595bd9fcf1f006a12f1338cb4480acea383b3b65da10bb9f28     0.002159
c904d4c3485db6569c9deb41da845d82056e8843eafba8bc8a90aaa06df96963     0.002159
59e19706d51d39f66711c2653cd7eb1291c94d9b55eb14bda74ce4dc636d015a     0.002159
Name: Agent_id, Length: 120, dtype: float64
LOST    93.365287
WON      6.634713
Name: status, dtype: float64
Low availability                 

In [24]:
# Exploring Lease

data['lease'].head(50)

0            Full Year Course Stay 40 - 44 weeks
1            Full Year Course Stay 40 - 44 weeks
2            Full Year Course Stay 40 - 44 weeks
3                                              0
4            Full Year Course Stay 40 - 44 weeks
5                                  semester-stay
6                                              0
7                                             51
8                    Semester Stay 20 - 24 weeks
9                                             51
10                                 semester-stay
11           Full Year Course Stay 40 - 44 weeks
12           Full Year Course Stay 40 - 44 weeks
13    Complete Education Year Stay 50 - 52 weeks
14                              full-course-stay
15                              full-course-stay
16                                full-year-stay
17                                         Other
18           Full Year Course Stay 40 - 44 weeks
19    Complete Education Year Stay 50 - 52 weeks
20           Full Ye

In [25]:
data['lease'] = data['lease'].str.lower()

In [26]:
data['lease'][data['lease'].str.contains('year')].value_counts()

full year course stay 40 - 44 weeks           13336
complete education year stay 50 - 52 weeks     6451
full-year-stay                                  956
1 year                                           90
full year stay                                   78
full year                                        27
academic year                                    10
1 year .                                          5
2 years                                           3
1year                                             3
full yyear stay                                   1
6m to 1 year                                      1
2years                                            1
1 year stay                                       1
full year course stay 44- 51 weeks                1
full year stayt                                   1
1 year stay.                                      1
3 years                                           1
full year course stay 51 weeks                    1
whole year  

In [27]:
data['lease'][data['lease'].str.contains('academic')].value_counts()

academic year    10
academic yr       1
Name: lease, dtype: int64

In [28]:
data['lease'][data['lease'].str.contains('education')].value_counts()

complete education year stay 50 - 52 weeks    6451
Name: lease, dtype: int64

In [29]:
data['lease'][data['lease'].str.contains('semester')].value_counts()

semester stay 20 - 24 weeks    2753
semester-stay                   470
semester stay                     5
Name: lease, dtype: int64

In [30]:
data['lease'] = data['lease'].replace(['full-year-stay', '1 year', 'full year stay', 'full year', '1 year .', '1year', 'full year course stay 40 - 44 weeks', 'full yyear stay', '1 year stay', 'full year course stay 44- 51 weeks', 'full year stayt', '1 year stay.', 'full year course stay 51 weeks', 'whole year'], 'full year')

In [31]:
data['lease'][data['lease'].str.contains('year')].value_counts()

full year                                     14502
complete education year stay 50 - 52 weeks     6451
academic year                                    10
2 years                                           3
3 years                                           1
2years                                            1
6m to 1 year                                      1
Name: lease, dtype: int64

In [32]:
data['lease'] = data['lease'].replace(['complete education year stay 50 - 52 weeks', 'academic year', 'academic yr'], 'academic year')

In [33]:
data['lease'][data['lease'].str.contains('academic')].value_counts()

academic year    6462
Name: lease, dtype: int64

In [34]:
data['lease'] = data['lease'].replace(['semester stay 20 - 24 weeks', 'semester-stay', 'semester stay'], 'semester stay')

In [35]:
data['lease'].value_counts(normalize=True)*100

full year            31.310318
0                    16.274802
academic year        13.951681
51                    8.532504
semester stay         6.969363
                       ...    
36-40 weeks           0.002159
12 month              0.002159
44 weeks/51 weeks     0.002159
25-30 weeks           0.002159
78                    0.002159
Name: lease, Length: 271, dtype: float64

In [36]:
# replace all values with less than 10% by other
data['lease'][(data['lease'] != 'full year') & (data['lease'] != '0') & (data['lease'] != '51') & (data['lease'] != 'academic year')] = 'other'

In [37]:
data['lease'].value_counts(normalize = True)*100

full year        31.310318
other            29.930695
0                16.274802
academic year    13.951681
51                8.532504
Name: lease, dtype: float64

In [38]:
data.columns

Index(['Agent_id', 'status', 'lost_reason', 'budget', 'lease', 'movein',
       'source', 'source_city', 'source_country', 'utm_source', 'utm_medium',
       'des_city', 'des_country', 'lead_id'],
      dtype='object')

In [39]:
data['Agent_id'].value_counts(normalize = True)*100

2fca346db656187102ce806ac732e06a62df0dbb2829e511a770556d398e1a6e    17.673856
96ab6152cb421b4d4012a4903c19004608a342a5f7c6671a7828b8ad8d5b9489     3.132759
44864c96fa1c36602f0d045b268981b6cab638a60fc207c89e3bba235f619450     2.761405
fb3e2c0015ad5bcd60d7619ca575a56c15c5d2cd3f01f18e68c4dfaeee585db5     2.385733
eac9815a500f908736d303e23aa227f0957177b0e6756b915a3d3cc3a846b83b     2.308008
                                                                      ...    
e16d12cd21f38bc8253f65cc5c0bf4932cf1397ba1c60682b2a99049e0e7ceed     0.002159
04222ea3e14cb1209b9726defe3efce5196b7afa0a959854a30401be41f4026d     0.002159
fb87e6cf60af08595bd9fcf1f006a12f1338cb4480acea383b3b65da10bb9f28     0.002159
c904d4c3485db6569c9deb41da845d82056e8843eafba8bc8a90aaa06df96963     0.002159
59e19706d51d39f66711c2653cd7eb1291c94d9b55eb14bda74ce4dc636d015a     0.002159
Name: Agent_id, Length: 120, dtype: float64

In [40]:
# replace all values with less than 10% by other
data['Agent_id'][(data['Agent_id'] != '2fca346db656187102ce806ac732e06a62df0dbb2829e511a770556d398e1a6e')] = 'other'

In [41]:
data['Agent_id'].value_counts(normalize = True)*100

other                                                               82.326144
2fca346db656187102ce806ac732e06a62df0dbb2829e511a770556d398e1a6e    17.673856
Name: Agent_id, dtype: float64

In [42]:
data['lost_reason'].value_counts(normalize=True)*100

Low availability                  25.832848
Not responding                    16.898763
Not interested                    13.923613
Low budget                        11.885485
Wants private accommodation        6.004275
Short stay                         4.888054
Not a student                      3.277414
Just Enquiring                     2.726861
Inadequate details                 1.785522
Semester stay                      1.763931
Booked with manager                1.459507
Junk lead                          1.358033
Booked on campus accommodation     1.323488
Supply issue                       1.161561
No supply                          1.152924
Not serving in that region         0.887363
Repeat lead                        0.850660
Booked with competitor             0.768616
Not going to university            0.710322
Distance issue                     0.472829
Junk lead/ Just Enquiring          0.317378
Lead issue                         0.274197
Looking for admission           

In [43]:
# replace all values with less than 10% by other
data['lost_reason'][(data['lost_reason'] != 'Low availability') & (data['lost_reason'] != 'Not responding') & (data['lost_reason'] != 'Not interested') & (data['lost_reason'] != 'Low budget')] = 'other'

In [44]:
data['lost_reason'].value_counts(normalize=True)*100

other               31.459291
Low availability    25.832848
Not responding      16.898763
Not interested      13.923613
Low budget          11.885485
Name: lost_reason, dtype: float64

In [45]:
data['movein'].value_counts(normalize=True)*100

10/09/22    34.674094
31/08/22     4.721808
01/09/22     4.572835
09/09/22     3.374571
17/09/22     3.070147
              ...    
26/04/23     0.002159
10/06/23     0.002159
05/01/24     0.002159
06/09/06     0.002159
06/06/23     0.002159
Name: movein, Length: 469, dtype: float64

In [46]:
# replace all values with less than 10% by other
data['movein'][data['movein'] != '10/09/22'] = 'other'

In [47]:
data['movein'].value_counts(normalize=True)*100

other       65.325906
10/09/22    34.674094
Name: movein, dtype: float64

In [48]:
data['source'].value_counts(normalize=True)*100

7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2b7c51be879f0c37ca8    54.068700
7bcfca0e9d73371699b0adbf1c691e02108fa64f02f4dbe24a0217f139a9b916     8.191377
ba2d0a29556ac20f86f45e4543c0825428cba33fd7a9eaa29e3f169d2ca43b2b     7.742298
146fb4ecbe78caa24102bbaac12e2559a8c8e32fb116d36e3553a709fce40549     4.771466
d684761c17c11590f6e2525b48141cb2c0c6f2be5df4e229dae06e64c5c41b64     4.596584
                                                                      ...    
ab4556b4118e9689282ac334ed94636124b8d4a531e2db6c2623eb0437de1ab2     0.002159
ff25c08fc834360514f4fbc6434c2bb381ad04ea984ea86c2a0f5fb6ec94a831     0.002159
fde7c38145e98695f302f2eb113737bcf9e6b3b0ec4ab2723f4d6610cc97813f     0.002159
f4dd29fa0ded5c9ca30c96015bf8147165bfce38964c3842352d7a013b870ac9     0.002159
93bee3e177d17c128e6a6af7b88fd675d7ec729b1a88ae13c582248dfd702378     0.002159
Name: source, Length: 682, dtype: float64

In [49]:
# replace all values with less than 10% by other
data['source'][data['source'] != '7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2b7c51be879f0c37ca8'] = 'other'

In [50]:
data['source'].value_counts(normalize=True)*100

7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2b7c51be879f0c37ca8    54.0687
other                                                               45.9313
Name: source, dtype: float64

In [51]:
data['source_city'].value_counts(normalize=True)*100

ecc0e7dc084f141b29479058967d0bc07dee25d9690a98ee4e6fdad5168274d7    25.586718
b384863fa1f6e091164b74219726eee0b9dd47776b91a4dc71fd0516630d21ec     3.089578
9f2ba6981e36ea0fca6c16f42e6413d788d2a7499b8b7f418c5d76f12d191f02     1.843815
7d1550b844ff586a6023216c06263105eed0a849a2a1f69bb8862ab288d8cdab     1.623594
810c069040f6a9b16fdf976a901755508a87cb0224b169d4b5d803357e647947     1.455189
                                                                      ...    
67e6f8c1f6f47e01ab6fe1cbbc3d4115d71f63c6ebf82a8713a6ac04a5b5236c     0.002159
c2cedb3b32de2ae419a258d76f4fa82ee302441fbae041acf9a13770dfa18eb0     0.002159
e2f36920bce1be5e4c94b165dcbd09cf454f5990d8869c5cc483e0723e34595c     0.002159
06065eeea31a9f7eeaf11e335f9a6a67235fc40f12d50b06b84fcbcb5ed70dd2     0.002159
d650065b53c2284b98e575d37e1be820599bf33baff5e147770547a984975d2f     0.002159
Name: source_city, Length: 4315, dtype: float64

In [52]:
# replace all values with less than 10% by other
data['source_city'][data['source_city'] != 'ecc0e7dc084f141b29479058967d0bc07dee25d9690a98ee4e6fdad5168274d7'] = 'other'

In [53]:
data['source_city'].value_counts(normalize=True)*100

other                                                               74.413282
ecc0e7dc084f141b29479058967d0bc07dee25d9690a98ee4e6fdad5168274d7    25.586718
Name: source_city, dtype: float64

In [54]:
data['source_country'].value_counts(normalize=True)*100

e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a9750aa698e797e9c96    58.917892
8da82000ef9c4468ba47362a924b895e40662fed846942a1870a674e5c6d1fc2    15.016085
e39b0c5e88f19053d3a917669bc9d60729f351e064ca0c94f5dc73f4e676333f     4.335341
38fe804a1f9ec032ad876bd7192c1f706e1402831e163cf72702b1f451f40cac     2.141762
0207c236c5ad89235d814b1e1807f6fdc1930810439489f76600e1672338e42b     0.995315
                                                                      ...    
0b6e4801d2a43e3c5dc06d5b68524bd9e80e425610b531d977b5713d59df47b6     0.002159
4a89aa60736d863a5fda8aa2305e299e1c3bea690fe4775363fbf2762102b273     0.002159
b828c41f8b20d533c028d8a4e1019fe0b16bb14d4b7114a662365d08e89a1bea     0.002159
39ae135c41c0e9f97bdbe7449f5320052c6a699cbf0fcbd95dc17f6465f0d34e     0.002159
282a5dbb9d8cb1ea0caf847b5aaca66cf1a76b5e57ddbbd2dd7db8204f553c5c     0.002159
Name: source_country, Length: 184, dtype: float64

In [55]:
# replace all values with less than 10% by other
data['source_country'][(data['source_country'] != 'e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a9750aa698e797e9c96') & (data['source_country'] != '8da82000ef9c4468ba47362a924b895e40662fed846942a1870a674e5c6d1fc2')] = 'other'

In [56]:
data['source_country'].value_counts(normalize=True)*100

e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a9750aa698e797e9c96    58.917892
other                                                               26.066023
8da82000ef9c4468ba47362a924b895e40662fed846942a1870a674e5c6d1fc2    15.016085
Name: source_country, dtype: float64

In [57]:
data['utm_source'].value_counts(normalize=True)*100

bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07ee6acc68762d12c2e3    66.787573
7f3fa48ca885678134842fa7456f3ece53a97f843b610185d900ac4e467c7490    10.259732
3d59f7548e1af2151b64135003ce63c0a484c26b9b8b166a7b1c1805ec34b00a     4.268411
3c77f261a156a5308fee53720276395ef78d2e7367e4225a3d3d93f4accd1dd3     3.758879
d15690f08a575024650b01ffac892cfd2b93e6c57c140f1b6d9e47753cabd579     3.294687
9ac56ef275bb33f0f931abd846e53e845a80af8a549100741e928b8b2abd56a7     2.698793
ec8202b6f9fb16f9e26b66367afa4e037752f3c09a18cefab426165e06a424b1     2.260509
4f48c17d2a97a7461a12d0d07336f808e70a1248a7082dbb103047f54090c158     1.865406
44574c4ba2ea74ad4bf1e184133cdbf4e7390a3690beff6a7364511a70ec208e     1.180992
4688fa9ce4f64d0451be91f30a3de24a1c1c41635bd221694b2a96f9ecd8d717     0.958611
5ede56738376a8c53a7c99bf275b077a5c10df780a20cf3d8282fd59a2a9fb56     0.509532
7edb360f06acaef2cc80dba16cf563f199d347db4443da04da0c8173e3f9e4ed     0.373513
3f40462915a3e6026a4d790127b95ded4d870f6ab18d9af2fcbc454168255237

In [58]:
# replace all values with less than 10% by other
data['utm_source'][(data['utm_source'] != 'bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07ee6acc68762d12c2e3') & (data['utm_source'] != '7f3fa48ca885678134842fa7456f3ece53a97f843b610185d900ac4e467c7490')] = 'other'

In [59]:
data['utm_source'].value_counts(normalize=True)*100

bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07ee6acc68762d12c2e3    66.787573
other                                                               22.952696
7f3fa48ca885678134842fa7456f3ece53a97f843b610185d900ac4e467c7490    10.259732
Name: utm_source, dtype: float64

In [60]:
data['utm_medium'].value_counts(normalize=True)*100

09076eb7665d1fb9389c7c4517fee0b00e43092eb34821b09b5730c41ebcc50c    74.888270
09bc8f0bb77bdddeb98527a39e995c2c605166399c178cb296992abab8415d5b     3.793424
69f81619d0ef92cbb165a44e76f4f0c284d2aa83c318fa7fb632207f68215647     3.577520
268ad70eb5bc4737a2ae28162cbca30118cc94520e49ef1ac5f72c85d3f2caa9     3.501954
82244417f956ac7c599f191593f7e441a4fafa20a4158fd52e154f1dc4c8ed92     2.431073
                                                                      ...    
974f37cbfab65a440154a8435e81f1928b6cf003b2885291e7c6fddd530ede65     0.002159
6a3aea19761ac8238e4e1d298f15207fe12375f13cb79dfba243aabed36163ca     0.002159
ad89c0825db17eb2c0b1a0d1a04f712b1c531ac84a6c92901a7f29e96adae9ce     0.002159
511f27cf310544befd3236d40f86ec47f539c4722370acde0f4df8b6616c8553     0.002159
5f5238f61f941641a13c52f68a37a1de3e3d981b57506b82aa36d5b85d09c8b7     0.002159
Name: utm_medium, Length: 63, dtype: float64

In [61]:
# replace all values with less than 10% by other
data['utm_medium'][(data['utm_medium'] != '09076eb7665d1fb9389c7c4517fee0b00e43092eb34821b09b5730c41ebcc50c')] = 'other'

In [62]:
data['utm_medium'].value_counts(normalize=True)*100

09076eb7665d1fb9389c7c4517fee0b00e43092eb34821b09b5730c41ebcc50c    74.88827
other                                                               25.11173
Name: utm_medium, dtype: float64

In [63]:
data['des_city'].value_counts(normalize=True)*100

ecc0e7dc084f141b29479058967d0bc07dee25d9690a98ee4e6fdad5168274d7    21.270808
810c069040f6a9b16fdf976a901755508a87cb0224b169d4b5d803357e647947     7.446510
11ab03a1a8c367191355c152f39fe28cae5e426fce49efb320230ca4ae3f97a1     6.703802
8593b9db65212160979d946950601c26622a219c80f1f122679eb69ec5b96600     4.333182
6a5ed83658ef85afc77709cc16ede854ad98e3c2ad8b076e10dcbcd6e5096271     3.668200
                                                                      ...    
6b9031df058346a877deaba0a830b576fa81b9da8692e183e4eb4c9dc949054c     0.002159
829228a55ddc193d4f48f1249126cd689365f5f7fe55383ead8afc2b8676de8c     0.002159
b9ae62ede2dad179198540d5a84bf5e432f8f36c370d906b9a8224d04582d9d0     0.002159
80aa149d3dedd651f836d1071659255aa4843801d79d5df18ddb8d5df1dcf32d     0.002159
806b1459b451fcbf14da3e1f61c2e0379e840b0a20badf7f6d4fdc7b3b9329b8     0.002159
Name: des_city, Length: 219, dtype: float64

In [64]:
# replace all values with less than 5% by other
data['des_city'][(data['des_city'] != 'ecc0e7dc084f141b29479058967d0bc07dee25d9690a98ee4e6fdad5168274d7') & (data['des_city'] != '810c069040f6a9b16fdf976a901755508a87cb0224b169d4b5d803357e647947') & (data['des_city'] != '11ab03a1a8c367191355c152f39fe28cae5e426fce49efb320230ca4ae3f97a1')] = 'other'

In [65]:
data['des_city'].value_counts(normalize=True)*100

other                                                               64.578880
ecc0e7dc084f141b29479058967d0bc07dee25d9690a98ee4e6fdad5168274d7    21.270808
810c069040f6a9b16fdf976a901755508a87cb0224b169d4b5d803357e647947     7.446510
11ab03a1a8c367191355c152f39fe28cae5e426fce49efb320230ca4ae3f97a1     6.703802
Name: des_city, dtype: float64

In [66]:
data['des_country'].value_counts(normalize=True)*100

8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512ff098eaa56a4d5fb19    96.983829
02bf1dfa9a0365a86223f0e4ac9eaa0517b06b2e9273790f719feda656a247ae     1.604163
80db4ccdca106d37b920206331fcfe3e9e50a9e763d89b54ce3ad5ac8cf30f03     0.349764
be55ef3f4c4e6c2d9c2afe2a33ac90ad0f50d4de7f9163999877e2a9ca5a54f8     0.321696
7a1ca4ef7515f7276bae7230545829c27810c9d9e98ab2c06066bee6270d5153     0.246130
c1ef40ce0484c698eb4bd27fe56c1e7b68d74f9780ed674210d0e5013dae45e9     0.166246
5a9cf672c8be6b5ab9546a2fb49b06dd81a4e364c86ed023898c49d9bb0605dc     0.123065
49dca65f362fee401292ed7ada96f96295eab1e589c52e4e66bf4aedda715fdd     0.056135
abd149214539d9f222d25de6358735b9fa0efd3956f66102b2c119ae2d9f6348     0.049658
8ef41e6f4b07432a0cb4eb7a8774e7a3878fd3e385f49aa09b406768467db228     0.047499
4c7de2c3da6dc0ae8f9d13b107718a913859359bd30166d077c867577953865d     0.036704
3a386c13d283dc2ebbe2aeffc1ab3930d086b28956930d913d8dc58f34c013b7     0.008636
2275583196d791405892aaca0d87743c872f3fc0cf3308a6c3ef82528918aa8a

In [67]:
# replace all values with less than 10% by other
data['des_country'][(data['des_country'] != '8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512ff098eaa56a4d5fb19')] = 'other'

In [68]:
data['des_country'].value_counts(normalize=True)*100

8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512ff098eaa56a4d5fb19    96.983829
other                                                                3.016171
Name: des_country, dtype: float64

In [69]:
#As lead id is the id for each lead we can drop the column
data.drop('lead_id', axis=1, inplace=True)

In [70]:
# Printing Normalized value counts of all columns

def value_count_all(data):
  for c in data.columns:
    print(data[c].value_counts(normalize=True)*100)

value_count_all(data)

other                                                               82.326144
2fca346db656187102ce806ac732e06a62df0dbb2829e511a770556d398e1a6e    17.673856
Name: Agent_id, dtype: float64
LOST    93.365287
WON      6.634713
Name: status, dtype: float64
other               31.459291
Low availability    25.832848
Not responding      16.898763
Not interested      13.923613
Low budget          11.885485
Name: lost_reason, dtype: float64
£60 - £120 Per week     29.026060
£121 - £180 Per Week    19.167908
0-0                     12.347518
£181 - £240 Per Week     4.626811
100                      3.257983
                          ...    
0-80-100$                0.002159
0-235                    0.002159
£120-160 Per Week        0.002159
1400 per month           0.002159
0-500 PCM                0.002159
Name: budget, Length: 1851, dtype: float64
full year        31.310318
other            29.930695
0                16.274802
academic year    13.951681
51                8.532504
Name: lease,

In [71]:
# dealing with budget
data['budget'].value_counts()

£60 - £120 Per week     13444
£121 - £180 Per Week     8878
0-0                      5719
£181 - £240 Per Week     2143
100                      1509
                        ...  
0-80-100$                   1
0-235                       1
£120-160 Per Week           1
1400 per month              1
0-500 PCM                   1
Name: budget, Length: 1851, dtype: int64

In [72]:
data['budget'][data['budget'].str.contains('£')]

0         £60 - £120 Per week
1         £60 - £120 Per week
2        £121 - £180 Per Week
4         £60 - £120 Per week
7        £181 - £240 Per Week
                 ...         
46602    £121 - £180 Per Week
46603     £60 - £120 Per week
46604     £60 - £120 Per week
46605    £241 - £300 Per week
46607    £181 - £240 Per Week
Name: budget, Length: 26276, dtype: object

In [73]:
data['budget'] = data['budget'].replace('pounds', '£')

In [74]:
(data['budget'][~ data['budget'].str.contains('£')].size / data.shape[0]) * 100

43.269210009283846

In [75]:
data['budget'].value_counts(normalize=True)*100

£60 - £120 Per week     29.026060
£121 - £180 Per Week    19.167908
0-0                     12.347518
£181 - £240 Per Week     4.626811
100                      3.257983
                          ...    
0-80-100$                0.002159
0-235                    0.002159
£120-160 Per Week        0.002159
1400 per month           0.002159
0-500 PCM                0.002159
Name: budget, Length: 1851, dtype: float64

In [76]:
data['budget'][data['budget'].str.contains('week')]

0         £60 - £120 Per week
1         £60 - £120 Per week
4         £60 - £120 Per week
17        £60 - £120 Per week
18        £60 - £120 Per week
                 ...         
46598     £60 - £120 Per week
46599     £60 - £120 Per week
46603     £60 - £120 Per week
46604     £60 - £120 Per week
46605    £241 - £300 Per week
Name: budget, Length: 14782, dtype: object

In [77]:
data['budget'] = data['budget'].str.lower()

In [78]:
data['budget'].value_counts(normalize=True)*100

£60 - £120 per week     29.028219
£121 - £180 per week    19.167908
0-0                     12.347518
£181 - £240 per week     4.626811
100                      3.257983
                          ...    
£120-160 per week        0.002159
1400 per month           0.002159
0-£80 - £90 per week     0.002159
500 euros month          0.002159
0-500 pcm                0.002159
Name: budget, Length: 1814, dtype: float64

In [79]:
# replace all values with less than 10% by other
data['budget'][(data['budget'] != '£60 - £120 per week') & (data['budget'] != '£121 - £180 per week') & (data['budget'] != '0-0')] = 'other'

In [80]:
data['budget'][data['budget'] == '0-0'] = '£0 - £0 per week'

In [81]:
data['budget'].value_counts(normalize=True)*100

other                   39.456355
£60 - £120 per week     29.028219
£121 - £180 per week    19.167908
£0 - £0 per week        12.347518
Name: budget, dtype: float64

In [82]:
data.describe().T

count unique  \
Agent_id        46317      2   
status          46317      2   
lost_reason     46317      5   
budget          46317      4   
lease           46317      5   
movein          46317      2   
source          46317      2   
source_city     46317      2   
source_country  46317      3   
utm_source      46317      3   
utm_medium      46317      2   
des_city        46317      4   
des_country     46317      2   

                                                              top   freq  
Agent_id                                                    other  38131  
status                                                       LOST  43244  
lost_reason                                                 other  14571  
budget                                                      other  18275  
lease                                                   full year  14502  
movein                                                      other  30257  
source          7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...  25043  
source_city                                                 other  34466  
source_country  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...  27289  
utm_source      bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07...  30934  
utm_medium      09076eb7665d1fb9389c7c4517fee0b00e43092eb34821...  34686  
des_city                                                    other  29911  
des_country     8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...  44920

In [83]:
missing_percentage()

column  missing_percentage
0         Agent_id                 0.0
1           status                 0.0
2      lost_reason                 0.0
3           budget                 0.0
4            lease                 0.0
5           movein                 0.0
6           source                 0.0
7      source_city                 0.0
8   source_country                 0.0
9       utm_source                 0.0
10      utm_medium                 0.0
11        des_city                 0.0
12     des_country                 0.0

In [84]:
data.to_csv('uniacco_cleaned.csv')

In [85]:
data.head()

Agent_id status     lost_reason                budget      lease    movein  \
0    other   LOST  Not responding   £60 - £120 per week  full year  10/09/22   
1    other   LOST      Low budget   £60 - £120 per week  full year  10/09/22   
2    other   LOST  Not responding  £121 - £180 per week  full year     other   
3    other   LOST      Low budget      £0 - £0 per week          0  10/09/22   
4    other   LOST           other   £60 - £120 per week  full year  10/09/22   

                                              source  \
0  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   
1  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   
2  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   
3                                              other   
4  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   

                                         source_city  \
0  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   
1  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   
2                                              other   
3                                              other   
4  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   

                                      source_country  \
0  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
1  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
2  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
3  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
4  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   

                                          utm_source  \
0                                              other   
1                                              other   
2  bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07...   
3  bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07...   
4                                              other   

                                          utm_medium  \
0                                              other   
1                                              other   
2  09076eb7665d1fb9389c7c4517fee0b00e43092eb34821...   
3  09076eb7665d1fb9389c7c4517fee0b00e43092eb34821...   
4                                              other   

                                            des_city  \
0  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   
1                                              other   
2  11ab03a1a8c367191355c152f39fe28cae5e426fce49ef...   
3                                              other   
4  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   

                                         des_country  
0  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...  
1  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...  
2  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...  
3  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...  
4  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...

In [86]:
X = data.drop('status', axis=1)
y = data.status
y = np.array([0 if i == 'LOST' else 1 for i in y])

In [87]:
X.head()

Agent_id     lost_reason                budget      lease    movein  \
0    other  Not responding   £60 - £120 per week  full year  10/09/22   
1    other      Low budget   £60 - £120 per week  full year  10/09/22   
2    other  Not responding  £121 - £180 per week  full year     other   
3    other      Low budget      £0 - £0 per week          0  10/09/22   
4    other           other   £60 - £120 per week  full year  10/09/22   

                                              source  \
0  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   
1  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   
2  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   
3                                              other   
4  7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2...   

                                         source_city  \
0  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   
1  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   
2                                              other   
3                                              other   
4  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   

                                      source_country  \
0  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
1  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
2  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
3  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   
4  e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a...   

                                          utm_source  \
0                                              other   
1                                              other   
2  bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07...   
3  bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07...   
4                                              other   

                                          utm_medium  \
0                                              other   
1                                              other   
2  09076eb7665d1fb9389c7c4517fee0b00e43092eb34821...   
3  09076eb7665d1fb9389c7c4517fee0b00e43092eb34821...   
4                                              other   

                                            des_city  \
0  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   
1                                              other   
2  11ab03a1a8c367191355c152f39fe28cae5e426fce49ef...   
3                                              other   
4  ecc0e7dc084f141b29479058967d0bc07dee25d9690a98...   

                                         des_country  
0  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...  
1  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...  
2  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...  
3  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...  
4  8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512...

In [88]:
y

array([0, 0, 0, ..., 0, 0, 0])

In [89]:
X_one_hot = pd.get_dummies(X)
X_one_hot.head()

Agent_id_2fca346db656187102ce806ac732e06a62df0dbb2829e511a770556d398e1a6e  \
0                                                  0                           
1                                                  0                           
2                                                  0                           
3                                                  0                           
4                                                  0                           

   Agent_id_other  lost_reason_Low availability  lost_reason_Low budget  \
0               1                             0                       0   
1               1                             0                       1   
2               1                             0                       0   
3               1                             0                       1   
4               1                             0                       0   

   lost_reason_Not interested  lost_reason_Not responding  lost_reason_other  \
0                           0                           1                  0   
1                           0                           0                  0   
2                           0                           1                  0   
3                           0                           0                  0   
4                           0                           0                  1   

   budget_other  budget_£0 - £0 per week  budget_£121 - £180 per week  ...  \
0             0                        0                            0  ...   
1             0                        0                            0  ...   
2             0                        0                            1  ...   
3             0                        1                            0  ...   
4             0                        0                            0  ...   

   utm_source_bbdefa2950f49882f295b1285d4fa9dec45fc4144bfb07ee6acc68762d12c2e3  \
0                                                  0                             
1                                                  0                             
2                                                  1                             
3                                                  1                             
4                                                  0                             

   utm_source_other  \
0                 1   
1                 1   
2                 0   
3                 0   
4                 1   

   utm_medium_09076eb7665d1fb9389c7c4517fee0b00e43092eb34821b09b5730c41ebcc50c  \
0                                                  0                             
1                                                  0                             
2                                                  1                             
3                                                  1                             
4                                                  0                             

   utm_medium_other  \
0                 1   
1                 1   
2                 0   
3                 0   
4                 1   

   des_city_11ab03a1a8c367191355c152f39fe28cae5e426fce49efb320230ca4ae3f97a1  \
0                                                  0                           
1                                                  0                           
2                                                  1                           
3                                                  0                           
4                                                  0                           

   des_city_810c069040f6a9b16fdf976a901755508a87cb0224b169d4b5d803357e647947  \
0                                                  0                           
1                                                  0                           
2                                                  0                           
3                                                  0    

In [90]:
X_one_hot.columns

Index(['Agent_id_2fca346db656187102ce806ac732e06a62df0dbb2829e511a770556d398e1a6e',
       'Agent_id_other', 'lost_reason_Low availability',
       'lost_reason_Low budget', 'lost_reason_Not interested',
       'lost_reason_Not responding', 'lost_reason_other', 'budget_other',
       'budget_£0 - £0 per week', 'budget_£121 - £180 per week',
       'budget_£60 - £120 per week', 'lease_0', 'lease_51',
       'lease_academic year', 'lease_full year', 'lease_other',
       'movein_10/09/22', 'movein_other',
       'source_7aae3e886e89fc1187a5c47d6cea1c22998ee610ade1f2b7c51be879f0c37ca8',
       'source_other',
       'source_city_ecc0e7dc084f141b29479058967d0bc07dee25d9690a98ee4e6fdad5168274d7',
       'source_city_other',
       'source_country_8da82000ef9c4468ba47362a924b895e40662fed846942a1870a674e5c6d1fc2',
       'source_country_e09e10e67812e9d236ad900e5d46b4308fc62f5d69446a9750aa698e797e9c96',
       'source_country_other',
       'utm_source_7f3fa48ca885678134842fa7456f3ece53a97f843

In [91]:
# Removing _other column for one-hot encoded columns to avoid dummy variable trap
X_one_hot.drop([
       'Agent_id_other',
       'lost_reason_other', 
       'budget_other',
       'lease_other',
       'movein_other',
       'source_other',
       'source_city_other',
       'source_country_other',
       'utm_source_other',
       'utm_medium_other',
       'des_city_other',
       'des_country_other'], axis=1, inplace=True)

In [92]:
X_one_hot.shape

(46317, 24)

### Feature Selection

In [93]:
# Feature Selection using Reccursive Feature Elemination using logistic regression model

from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression

logModel = LogisticRegression()
rfe = RFE(logModel)
fit = rfe.fit(X_one_hot, y)
print("Num Features: %d" % fit.n_features_)
print("Selected Features: %s" % fit.support_)
print("Feature Ranking: %s" % fit.ranking_)

Num Features: 12
Selected Features: [ True  True  True  True  True False False False  True  True False False
 False False False False False False False  True  True  True  True  True]
Feature Ranking: [ 1  1  1  1  1  5 13 11  1  1  9  8  7  2 10  6  4  3 12  1  1  1  1  1]


In [94]:
selected_cols = X_one_hot.columns[fit.support_]

In [95]:
selected_cols

Index(['Agent_id_2fca346db656187102ce806ac732e06a62df0dbb2829e511a770556d398e1a6e',
       'lost_reason_Low availability', 'lost_reason_Low budget',
       'lost_reason_Not interested', 'lost_reason_Not responding', 'lease_0',
       'lease_51',
       'utm_medium_09076eb7665d1fb9389c7c4517fee0b00e43092eb34821b09b5730c41ebcc50c',
       'des_city_11ab03a1a8c367191355c152f39fe28cae5e426fce49efb320230ca4ae3f97a1',
       'des_city_810c069040f6a9b16fdf976a901755508a87cb0224b169d4b5d803357e647947',
       'des_city_ecc0e7dc084f141b29479058967d0bc07dee25d9690a98ee4e6fdad5168274d7',
       'des_country_8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512ff098eaa56a4d5fb19'],
      dtype='object')

In [96]:
X_fs = X_one_hot[selected_cols]

In [97]:
X_fs.head()

Agent_id_2fca346db656187102ce806ac732e06a62df0dbb2829e511a770556d398e1a6e  \
0                                                  0                           
1                                                  0                           
2                                                  0                           
3                                                  0                           
4                                                  0                           

   lost_reason_Low availability  lost_reason_Low budget  \
0                             0                       0   
1                             0                       1   
2                             0                       0   
3                             0                       1   
4                             0                       0   

   lost_reason_Not interested  lost_reason_Not responding  lease_0  lease_51  \
0                           0                           1        0         0   
1                           0                           0        0         0   
2                           0                           1        0         0   
3                           0                           0        1         0   
4                           0                           0        0         0   

   utm_medium_09076eb7665d1fb9389c7c4517fee0b00e43092eb34821b09b5730c41ebcc50c  \
0                                                  0                             
1                                                  0                             
2                                                  1                             
3                                                  1                             
4                                                  0                             

   des_city_11ab03a1a8c367191355c152f39fe28cae5e426fce49efb320230ca4ae3f97a1  \
0                                                  0                           
1                                                  0                           
2                                                  1                           
3                                                  0                           
4                                                  0                           

   des_city_810c069040f6a9b16fdf976a901755508a87cb0224b169d4b5d803357e647947  \
0                                                  0                           
1                                                  0                           
2                                                  0                           
3                                                  0                           
4                                                  0                           

   des_city_ecc0e7dc084f141b29479058967d0bc07dee25d9690a98ee4e6fdad5168274d7  \
0                                                  1                           
1                                                  0                           
2                                                  0                           
3                                                  0                           
4                                                  1                           

   des_country_8d23a6e37e0a6431a8f1b43a91026dcff51170a89a6512ff098eaa56a4d5fb19  
0                                                  1                             
1                                                  1                             
2                                                  1                             
3                                                  1                             
4                                                  1

## Model Training

In [98]:
eval = {"model":[], "acc":[], "precision":[], "recall":[], "f1":[]}

In [99]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_fs, y, test_size=0.2, random_state=42)

In [100]:
X_train.shape, y_train.shape

((37053, 12), (37053,))

In [101]:
X_test.shape, y_test.shape

((9264, 12), (9264,))

### Logistic Regression

In [102]:
from sklearn.linear_model import LogisticRegression

model1 = LogisticRegression()
model1.fit(X_train, y_train)
y_pred = model1.predict(X_test)

In [103]:
y_pred

array([0, 1, 0, ..., 0, 0, 0])

In [104]:
y_test

array([0, 0, 0, ..., 0, 0, 0])

In [105]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

def get_metrics(y_pred, y_test):
  acc = accuracy_score(y_pred, y_test)
  precision = precision_score(y_pred, y_test)
  recall = recall_score(y_pred, y_test)
  f1 = f1_score(y_pred, y_test)

  return acc, precision, recall, f1

In [106]:
acc, precision, recall, f1 = get_metrics(y_pred, y_test)

In [107]:
eval["model"].append('logistic regression')
eval["acc"].append(acc)
eval["precision"].append(precision)
eval["recall"].append(recall)
eval["f1"].append(f1)
eval

{'model': ['logistic regression'],
 'acc': [0.9810017271157168],
 'precision': [0.915807560137457],
 'recall': [0.8075757575757576],
 'f1': [0.8582930756843802]}

### Decision Tree

In [108]:
from sklearn.tree import DecisionTreeClassifier

tree = DecisionTreeClassifier()
tree.fit(X_train, y_train)
y_pred = tree.predict(X_test)

In [109]:
acc, precision, recall, f1 = get_metrics(y_pred, y_test)
eval["model"].append('decision tree')
eval["acc"].append(acc)
eval["precision"].append(precision)
eval["recall"].append(recall)
eval["f1"].append(f1)
eval

{'model': ['logistic regression', 'decision tree'],
 'acc': [0.9810017271157168, 0.9819732297063903],
 'precision': [0.915807560137457, 0.9398625429553265],
 'recall': [0.8075757575757576, 0.8055964653902798],
 'f1': [0.8582930756843802, 0.8675654242664552]}

### Random Forest

In [110]:
from sklearn.ensemble import RandomForestClassifier

forest = RandomForestClassifier(n_estimators=100)
forest.fit(X_train, y_train)
y_pred = forest.predict(X_test)

In [111]:
acc, precision, recall, f1 = get_metrics(y_pred, y_test)
eval["model"].append('Random Forest')
eval["acc"].append(acc)
eval["precision"].append(precision)
eval["recall"].append(recall)
eval["f1"].append(f1)
eval

{'model': ['logistic regression', 'decision tree', 'Random Forest'],
 'acc': [0.9810017271157168, 0.9819732297063903, 0.9819732297063903],
 'precision': [0.915807560137457, 0.9398625429553265, 0.9398625429553265],
 'recall': [0.8075757575757576, 0.8055964653902798, 0.8055964653902798],
 'f1': [0.8582930756843802, 0.8675654242664552, 0.8675654242664552]}

### SVM

In [112]:
from sklearn.svm import SVC

svm = SVC()
svm.fit(X_train, y_train)
y_pred = svm.predict(X_test)

In [113]:
acc, precision, recall, f1 = get_metrics(y_pred, y_test)
eval["model"].append('Support Vector Classifier')
eval["acc"].append(acc)
eval["precision"].append(precision)
eval["recall"].append(recall)
eval["f1"].append(f1)
eval

{'model': ['logistic regression',
  'decision tree',
  'Random Forest',
  'Support Vector Classifier'],
 'acc': [0.9810017271157168,
  0.9819732297063903,
  0.9819732297063903,
  0.9819732297063903],
 'precision': [0.915807560137457,
  0.9398625429553265,
  0.9398625429553265,
  0.9398625429553265],
 'recall': [0.8075757575757576,
  0.8055964653902798,
  0.8055964653902798,
  0.8055964653902798],
 'f1': [0.8582930756843802,
  0.8675654242664552,
  0.8675654242664552,
  0.8675654242664552]}

### KNN

In [114]:
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)

In [115]:
acc, precision, recall, f1 = get_metrics(y_pred, y_test)
eval["model"].append('KNN')
eval["acc"].append(acc)
eval["precision"].append(precision)
eval["recall"].append(recall)
eval["f1"].append(f1)
eval

{'model': ['logistic regression',
  'decision tree',
  'Random Forest',
  'Support Vector Classifier',
  'KNN'],
 'acc': [0.9810017271157168,
  0.9819732297063903,
  0.9819732297063903,
  0.9819732297063903,
  0.9806778929188256],
 'precision': [0.915807560137457,
  0.9398625429553265,
  0.9398625429553265,
  0.9398625429553265,
  0.9106529209621993],
 'recall': [0.8075757575757576,
  0.8055964653902798,
  0.8055964653902798,
  0.8055964653902798,
  0.806697108066971],
 'f1': [0.8582930756843802,
  0.8675654242664552,
  0.8675654242664552,
  0.8675654242664552,
  0.8555286521388217]}

In [116]:
eval_df = pd.DataFrame(eval)
eval_df

model       acc  precision    recall        f1
0        logistic regression  0.981002   0.915808  0.807576  0.858293
1              decision tree  0.981973   0.939863  0.805596  0.867565
2              Random Forest  0.981973   0.939863  0.805596  0.867565
3  Support Vector Classifier  0.981973   0.939863  0.805596  0.867565
4                        KNN  0.980678   0.910653  0.806697  0.855529

## Model Evaluation

We can see that Decision Tree, Random Forest and SVM have the best metrics:
- Accuracy = 98.1973%
- Precision Score = 93.9863%
- Recall Score = 80.5596%
- F1-Score = 86.7565%